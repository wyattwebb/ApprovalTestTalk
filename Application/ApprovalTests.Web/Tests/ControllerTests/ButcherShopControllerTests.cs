using System.Collections.Generic;
using System.Linq;
using ApprovalTests.Reporters;
using ApprovalTests.Web.Controllers;
using ApprovalTests.Web.Models.BaconViewModels;
using ApprovalTests.Web.PersistanceModels.BaconModels;
using ApprovalTests.Web.Services;
using ApprovalTests.Web.Tests.Setup;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using Moq;
using Ploeh.AutoFixture;

namespace ApprovalTests.Web.Tests.ControllerTests
{
    [TestClass]
    public class ButcherShopControllerTests : ApiTestBase
    {

        private readonly Fixture _fixture;

        public ButcherShopControllerTests()
        {
            _fixture = new Fixture();
            _fixture.Customize(new ApprovalTestCustomization());
        }

        protected override string BaseUrl
        {
            get { return "http://localhost:8000/api/"; }
        }

        [TestMethod]
        public void GET_GetPigs__Given_10Pigs__When_Mapper_Succeeds__Then_Expected_Mapped_Result()
        {
            // Arrange
            var count = 10;
            var pigs = _fixture.CreateMany<Pig>(10).ToArray();

            var turkeyBaconService = new Mock<IBaconService>();
            turkeyBaconService.Setup(x => x.GetGeneratedPigs(count))
                .Returns(() => pigs);

            var subject = new ButcherShopController(
                baconService: turkeyBaconService.Object);

            // Act
            var result = subject.GetPigs(count);

            // Assert
            Assert.AreEqual(pigs[0].Belly, result.Pigs[0].Belly);
            Assert.AreEqual(pigs[0].Back, result.Pigs[0].Back);
            Assert.AreEqual(pigs[0].Cannon, result.Pigs[0].Cannon);
            Assert.AreEqual(pigs[0].Cheek, result.Pigs[0].Cheek);
            Assert.AreEqual(pigs[0].Coffin, result.Pigs[0].Coffin);
            Assert.AreEqual(pigs[0].Crops, result.Pigs[0].Crops);
            Assert.AreEqual(pigs[0].Dewclaw, result.Pigs[0].Dewclaw);
            Assert.AreEqual(pigs[0].Ear, result.Pigs[0].Ear);
            // ... Assert the rest
            Assert.AreEqual(pigs[1].Belly, result.Pigs[0].Belly);
            Assert.AreEqual(pigs[1].Back, result.Pigs[0].Back);
            Assert.AreEqual(pigs[1].Cannon, result.Pigs[0].Cannon);
            Assert.AreEqual(pigs[1].Cheek, result.Pigs[0].Cheek);
            Assert.AreEqual(pigs[1].Coffin, result.Pigs[0].Coffin);
            Assert.AreEqual(pigs[1].Crops, result.Pigs[0].Crops);
            Assert.AreEqual(pigs[1].Dewclaw, result.Pigs[0].Dewclaw);
            Assert.AreEqual(pigs[1].Ear, result.Pigs[0].Ear);
            // ... Assert the rest up till 10

        }

        [TestMethod, Description("Approval Tests Assert")]
        public void GET_GetPigs_Ten()
        {
            // Arrange
            var count = 10;
            var pigs = _fixture.CreateMany<Pig>(10).ToArray();

            var turkeyBaconService = new Mock<IBaconService>();
            turkeyBaconService.Setup(x => x.GetGeneratedPigs(count))
                .Returns(() => pigs);

            var subject = new ButcherShopController(
                baconService: turkeyBaconService.Object);
            
            // Act
            var result = subject.GetPigs(count);

            // Assert
            VerifyApprovedJsonResult(result);
        }

        [TestMethod, Description("No AutoFixture")]
        public void GET_Given_Manual_Setups__When_GetPigs_Three__ThenAsserts()
        {
            // Arrange
            var count = 3;
            var pigs = new[]
            {
                new Pig
                {
                    Back = "Back",
                    Belly = "Belly",
                    Cannon = "Cannon",
                    Cheek = "Cheek",
                    Coffin = "Coffin",
                    Crops = "Crops",
                    Dewclaw = "Dewclaw",
                    Ear = "Ear",
                    FetLock = "FetLock",
                    ForeFlank = "ForFlank"
                    // TODO the rest..
                },
                new Pig
                {
                    Back = "Back",
                    Belly = "Belly",
                    Cannon = "Cannon",
                    Cheek = "Cheek",
                    Coffin = "Coffin",
                    Crops = "Crops",
                    Dewclaw = "Dewclaw",
                    Ear = "Ear",
                    FetLock = "FetLock",
                    ForeFlank = "ForFlank"
                    // TODO the rest..
                },
                new Pig
                {
                    Back = "Back",
                    Belly = "Belly",
                    Cannon = "Cannon",
                    Cheek = "Cheek",
                    Coffin = "Coffin",
                    Crops = "Crops",
                    Dewclaw = "Dewclaw",
                    Ear = "Ear",
                    FetLock = "FetLock",
                    ForeFlank = "ForFlank"
                    // TODO the rest..
                }
            };

            var turkeyBaconService = new Mock<IBaconService>();
            turkeyBaconService.Setup(x => x.GetGeneratedPigs(count))
                .Returns(() => pigs);

            var subject = new ButcherShopController(
                baconService: turkeyBaconService.Object);

            // Act
            var result = subject.GetPigs(count);

            // Assert
            VerifyApprovedJsonResult(result);
        }

        [TestMethod, Description("Approval Tests Assert With AutoFixture")]
        public void GET_GetPigs_Ten_UsingAutoFixture()
        {
            // Arrange
            var count = 10;
            var pigs = _fixture.CreateMany<Pig>(10).ToArray();

            var turkeyBaconService = new Mock<IBaconService>();
            turkeyBaconService.Setup(x => x.GetGeneratedPigs(count))
                .Returns(() => pigs);

            var subject = new ButcherShopController(
                baconService: turkeyBaconService.Object);

            // Act
            var result = subject.GetPigs(count);

            // Assert
            VerifyApprovedJsonResult(result);
        }

        [TestMethod, Description("Manual Asserts on null Validation")]
        public void GET_GetPigs_Given_Null_Parameter__When_Executed__Then_Return_Validation_Error_ManualAssertion()
        {
            // Arrange
            int? count = null;

            var expected = "Count is Required";
            var mockValidation = new Mock<IValidateInput>();
            mockValidation.Setup(x => x.ValidateGetPigs(count))
                .Returns(expected);

            var subject = new ButcherShopController(
                validateInput: mockValidation.Object);

            // Act
            var result = subject.GetPigs(count);

            // Assert
            Assert.AreEqual(result.ErrorMessage, expected);
        }

        [TestMethod, Description("Manual Asserts on min Validation")]
        public void GET_GetPigs_Given_Min_Parameter__When_Executed__Then_Return_Validation_Error_ManualAssertion()
        {
            // Arrange
            int? count = 5;

            var expected = "FIVE is OUTRIGHT!";
            var mockValidation = new Mock<IValidateInput>();
            mockValidation.Setup(x => x.ValidateGetPigs(count))
                .Returns(expected);

            var subject = new ButcherShopController(
                validateInput: mockValidation.Object);

            // Act
            var result = subject.GetPigs(count);

            // Assert
            Assert.AreEqual(result.ErrorMessage, expected);
        }

        [TestMethod, Description("Multiple Approval Asserts")]
        public void GET_GetPigs_Given_Parameter__When_Executed__Then_Return_Validation_Errors()
        {
            // Arrange
            int? nullCount = null;
            var fiveCount = 5;

            var nullExpected = "Count is Required";
            var mockValidation = new Mock<IValidateInput>();
            mockValidation.Setup(x => x.ValidateGetPigs(nullCount))
                .Returns(nullExpected);

            var fiveExpected = "FIVE is OUTRIGHT!";
            mockValidation.Setup(x => x.ValidateGetPigs(fiveCount))
                .Returns(fiveExpected);

            var subject = new ButcherShopController(
                validateInput: mockValidation.Object);
            var results = new List<PigsViewModel>();

            // Act
            results.Add(subject.GetPigs(nullCount));
            results.Add(subject.GetPigs(fiveCount));

            // Assert
            VerifyApprovedJsonResult(results);
        }

        [TestMethod, Description("Multiple Approval Asserts HTML")]
        public void GET_GetPigs_Given_Parameter__When_Executed__Then_Return_Validation_Errors_Html()
        {
            // Arrange
            int? nullCount = null;
            var fiveCount = 5;

            var nullExpected = "Count is Required";
            var mockValidation = new Mock<IValidateInput>();
            mockValidation.Setup(x => x.ValidateGetPigs(nullCount))
                .Returns(nullExpected);

            var fiveExpected = "FIVE is OUTRIGHT!";
            mockValidation.Setup(x => x.ValidateGetPigs(fiveCount))
                .Returns(fiveExpected);

            var subject = new ButcherShopController(
                validateInput: mockValidation.Object);
            var results = new List<PigsViewModel>();

            // Act
            var htmlReport = new BuildHtml("Multiple Approval Test Count Input");
            htmlReport.AddLine(subject.GetPigs(nullCount).ErrorMessage);
            htmlReport.AddLine(subject.GetPigs(fiveCount).ErrorMessage);

            // Assert
            VerifyApprovedHtmlResult(htmlReport.GetHtml());
        }

        [TestMethod]
        [UseReporter(typeof(DiffReporter))]
        public void GET_GetPigs_API_JsonTest()
        {
            var result = ExecuteGetRequest<PigsViewModel>("ButcherShop/GetPigs?id=1");

            VerifyApprovedJsonResult(result);
        }

    }
}
