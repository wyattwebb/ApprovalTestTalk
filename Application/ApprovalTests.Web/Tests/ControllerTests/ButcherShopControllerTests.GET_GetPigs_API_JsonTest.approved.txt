{
  "ErrorMessage": null,
  "Pigs": [
    {
      "Back": "Back84bbc7a3-6c45-4dfa-8220-7b68fd2f1e30",
      "Belly": "Bellyf10e7ce2-34b9-49ad-bb2a-a7f59ecb4081",
      "Cannon": "Cannon2c902015-53e7-4bd2-a308-7a68ca62c18f",
      "Cheek": "Cheek21c41261-bfe6-4980-ba71-0bffe009e5d4",
      "Coffin": "Coffinf291cd65-802b-4e9d-8678-e2547ae5ba45",
      "Crops": "Crops9da78df8-e1dd-4eda-be40-69588af35c3d",
      "Dewclaw": "Dewclaw06f084ae-d348-495a-9298-c3203452a428",
      "Ear": "Ear74941ec6-59c5-448a-9d94-098f516c13a3",
      "FetLock": "FetLockbc2af098-cc42-499d-ac32-d681be9c949c",
      "ForeFlank": "ForeFlank6573220d-89d6-4cb5-8875-0500b064252d",
      "ForeLeg": "ForeLeg9ab7bed2-f463-41de-bdd6-470415ff1916",
      "Ham": "Hame6474952-5576-41c4-8b54-b40897acf70b",
      "Hock": "Hockfc5ed13e-9ce5-4344-982f-8d5ee1cb9c8e",
      "Jowls": "Jowlsc4b24d88-47cb-40ad-a0ab-182e067ab27b",
      "Knee": "Kneef467b2b0-d8e2-4dbc-a3e3-560534319fb1",
      "Loin": "Loin85304de6-4735-4226-86e6-b47c46b1f03a",
      "Neck": "Neck139cefa8-4cdb-4454-a877-c95161cbd816",
      "Nostrils": "Nostrilsd47f8fea-183b-4983-a44d-7920d670a2c6",
      "Pastern": "Pastern929f1f2d-27e0-41dd-be7b-4b836d4dbda5",
      "Poll": "Polleea51939-2d29-4149-ad6f-6422f27acb02",
      "RearFlank": "RearFlankea8a5240-c809-4290-996b-e3a1f64412c0",
      "Rump": "Rump9cb65b1b-3158-4a4d-8dfb-a873baeb0bbe",
      "Sheath": "Sheathce5020d1-c63f-431d-b879-f5ee9290e185",
      "Shoulder": "Shoulder282babd2-4cfe-4143-b283-4c0dad1eae6a",
      "Snout": "Snout0a1a75e9-03dc-4a96-b9fd-35f773908fb2",
      "Stiffle": "Stiffle02255f29-710e-4541-bcc9-e08f0db1cd6f",
      "Tail": "Tail2154309d-eb84-41c8-9d15-92eaac9d1c53"
    }
  ]
}